Sources/HttpPipeline/Middleware.swift

```diff
@@ -15,90 +15,85 @@ public func writeStatus<A>(_ status: Status) -> Middleware<StatusLineOpen, Heade
+public func writeHeaders<E, A>(_ headers: [ResponseHeader]) -> Middleware<HeadersOpen, HeadersOpen, E, E, A, A> {
   return pure <<< (\.response.headers %~ { $0 + headers })
 }
 
-public func writeHeader<A>(_ name: String, _ value: String) -> Middleware<HeadersOpen, HeadersOpen, A, A> {
+public func writeHeader<E, A>(_ name: String, _ value: String) -> Middleware<HeadersOpen, HeadersOpen, E, E, A, A> {
   return writeHeader(.other(name, value))
 }
 
-public func closeHeaders<A>(conn: Conn<HeadersOpen, A>) -> IO<Conn<BodyOpen, A>> {
+public func closeHeaders<E, A>(conn: Conn<HeadersOpen, E, A>) -> IO<Conn<BodyOpen, E, A>> {
   return pure <| .init(
     data: conn.data,
     request: conn.request,
     response: conn.response
   )
 }
 
-public func end(conn: Conn<BodyOpen, Data>) -> IO<Conn<ResponseEnded, Data>> {
-  return pure <| .init(
-    data: conn.data,
-    request: conn.request,
-    response: Response(
-      status: conn.response.status,
-      headers: conn.response.headers,
-      body: conn.data
+public func end<E>(conn: Conn<BodyOpen, E, Data>) -> IO<Conn<ResponseEnded, E, Data>> {
+  switch conn.data {
+  case .left:
+    return pure <| .init(
+      data: conn.data,
+      request: conn.request,
+      response: conn.response
     )
-  )
+  case let .right(data):
+    return pure <| .init(
+      data: conn.data,
+      request: conn.request,
+      response: .init(
+        status: conn.response.status,
+        headers: conn.response.headers,
+        body: data
+      )
+    )
+  }
 }
 
-// TODO: rename to ignoreBody
-public func end<A>(conn: Conn<HeadersOpen, A>) -> IO<Conn<ResponseEnded, Data>> {
+public func ignoreBody<E, A>(conn: Conn<HeadersOpen, E, A>) -> IO<Conn<ResponseEnded, E, Data>> {
   return conn
     |> closeHeaders
     >-> map(const(Data())) >>> pure
     >-> end
 }
 
-public func redirect<A>(
+public func redirect<E, A>(
   to location: String,
-  headersMiddleware: @escaping Middleware<HeadersOpen, HeadersOpen, A, A> = (id >>> pure)
+  headersMiddleware: @escaping Middleware<HeadersOpen, HeadersOpen, E, E, A, A> = (id >>> pure)
   )
   ->
-  Middleware<StatusLineOpen, ResponseEnded, A, Data> {
+  Middleware<StatusLineOpen, ResponseEnded, E, E, A, Data> {
 
     return writeStatus(.found)
       >-> headersMiddleware
       >-> writeHeader(.location(location))
-      >-> end
+      >-> ignoreBody
 }
 
-public func send(_ data: Data) -> Middleware<BodyOpen, BodyOpen, Data, Data> {
-  return { conn in
-
-    let concatenatedData = conn.data + data
-
-    return pure <| .init(
-      data: concatenatedData,
-      request: conn.request,
-      response: Response(
-        status: conn.response.status,
-        headers: conn.response.headers,
-        body: concatenatedData
-      )
-    )
-  }
+public func send<E>(_ data: Data) -> Middleware<BodyOpen, BodyOpen, E, E, Data, Data> {
+  return pure <<< map { $0 + data }
```

Comment:
This one's actually pretty simple, though it would need an overhaul if we ever add some kind of continuation support for streaming endpoints.